src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+                    public static PropertyFetch FetcherForProperty(PropertyInfo propertyInfo)
+                    {
+                        if (propertyInfo == null)
+                            return new PropertyFetch();     // returns null on any fetch.
+
+                        var typedPropertyFetcher = typeof(TypedFetchProperty<,>);
+                        var instantiatedTypedPropertyFetcher = typedPropertyFetcher.GetTypeInfo().MakeGenericType(
+                            propertyInfo.DeclaringType, propertyInfo.PropertyType);
+                        return (PropertyFetch)Activator.CreateInstance(instantiatedTypedPropertyFetcher, propertyInfo);
+                    }
+
+                    /// <summary>
+                    /// Given an object, fetch the property that this propertyFech represents. 
+                    /// </summary>
+                    public virtual object Fetch(object obj) { return null; }
+
+                    #region private 
+
+                    class TypedFetchProperty<Tobject, Tproperty> : PropertyFetch
+                    {
+                        public TypedFetchProperty(PropertyInfo property)
+                        {
+                            _propertyFetch = (Func<Tobject, Tproperty>)property.GetMethod.CreateDelegate(typeof(Func<Tobject, Tproperty>));
+                        }
+                        public override object Fetch(object obj)
+                        {
+                            return _propertyFetch((Tobject)obj);
+                        }
+                        private Func<Tobject, Tproperty> _propertyFetch;
+                    }
+                    #endregion
+                }
+
+                string _propertyName;
+                Type _expectedType;
+                PropertyFetch _fetchForExpectedType;
+                #endregion
+            }
+
+            string _outputName;
+            PropertySpec _fetches;
+            #endregion
+        }
+
+        /// <summary>
+        /// CallbackObserver is a adapter class that creats an observer (which you can pass
+        /// to IObservable.Subscribe), and calls the given callback every time the 'next' 
+        /// operation on the IObserver happens. 
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        internal class CallbackObserver<T> : IObserver<T>
+        {
+            public CallbackObserver(Action<T> callback) { _callback = callback; }
+
+            #region private 
+            public void OnCompleted() { }
+            public void OnError(Exception error) { }
+            public void OnNext(T value) { _callback(value); }
+
+            Action<T> _callback;
+            #endregion
+        }
+
+        // A linked list of IObservable subscriptions (which are IDisposable).  
+        // We use this to keep track of the DiagnosticSource subscriptions.  
+        // We use this linked list for thread atomicity and to avoid taking a 
+        // dependency on collections.  
```

Comment:
> to avoid taking a dependency on collections

Yet ```"System.Collections": "4.0.0"``` is being added to the project.json... am I'm misunderstanding?